### PR TITLE
Implement basic TLV functionality

### PR DESCRIPTION
**Pull request by nkohen**
Introduced basic TLV types (Ping, Pong, Error, Unknown)

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala [hunk 1]
```diff
@@ -0,0 +1,140 @@
+package org.bitcoins.core.protocol.tlv
+
+import org.bitcoins.core.number.UInt16
+import org.bitcoins.core.protocol.BigSizeUInt
+import org.bitcoins.crypto.{Factory, NetworkElement}
+import scodec.bits.ByteVector
+
+sealed trait TLV extends NetworkElement {
+  def tpe: BigSizeUInt
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala:10]: type

File: core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  override def fromTLVValue(value: ByteVector): PongTLV = {
+    val numIgnored = UInt16(value.take(2))
+    val ignored = value.drop(2).take(numIgnored.toLong)
+
+    PongTLV.forIgnored(ignored)
+  }
+
+  def forIgnored(ignored: ByteVector): PongTLV = {
+    new PongTLV(ignored)
+  }
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala:147]: Are there more types we should define?

File: core-test/src/test/scala/org/bitcoins/core/number/UInt16Test.scala [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        val result = num1.toInt + num2.toInt
+        if (result <= UInt16.max.toInt) {
+          assert(num1 + num2 == UInt16(result.toInt))
+        } else {
+          assert(Try(num1 + num2).isFailure)
+        }
+    }
+  }
+
+  it must "subtract zero correctly" in {
```
Comment (benthecarman) [inline core-test/src/test/scala/org/bitcoins/core/number/UInt16Test.scala:108]: maybe add a test that it fails to subtracting from 0 
ie `UInt16.zero - UInt16(10)`

File: core-test/src/test/scala/org/bitcoins/core/protocol/tlv/TLVTest.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    forAll(TLVGen.pingTLV) { ping =>
+      assert(PingTLV(ping.bytes) == ping)
+    }
+  }
+
+  "PongTLV" must "have serialization symmetry" in {
+    forAll(TLVGen.pongTLV) { pong =>
+      assert(PongTLV(pong.bytes) == pong)
+    }
+  }
```
Comment (benthecarman) [inline core-test/src/test/scala/org/bitcoins/core/protocol/tlv/TLVTest.scala:43]: Should add tests that a `PongTLV` has an `Pong` type, etc

File: core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+    require(
+      bytes.length >= prefixSize + length.num.toLong,
+      s"Length specified was $length but not enough bytes in ${bytes.drop(prefixSize)}")
+
+    val value = bytes.drop(prefixSize).take(length.num.toLong)
+
+    (tpe, length, value)
+  }
+
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala:42]: It'd be nice to have a list of `knownTypes`

**Review by benthecarman — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala (hunk 1 above)
Comment (nkohen) [reply core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala:10]: reserved word, in scala the convention is to use `tpe` for this reason

File: core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala (hunk 2 above)
Comment (nkohen) [reply core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala:147]: Do you mean for TLV? We can certainly add more later but these are more or less just to test things out and define conventions.

File: core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala (hunk 1 above)
Comment (benthecarman) [reply core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala:10]: Ah yeah, forgot

**Review by benthecarman — APPROVED** [review]

File: core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala (hunk 1 above)
Comment (rorp) [reply core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala:10]: &#96;type&#96;

Comment (rorp) [reply core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala:10]: or tlvType

File: core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  override def bytes: ByteVector = {
+    tpe.bytes ++ length.bytes ++ value
+  }
+}
+
+object TLV extends Factory[TLV] {
+
+  def decodeTLV(bytes: ByteVector): (BigSizeUInt, BigSizeUInt, ByteVector) = {
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala]: I would say we should just wrap this in a return type with a nice name and fields

I.e. `DecodeTLVResult` or something like that.

**Pull request merged**